File: Java/interview_38.java
class Solution {
    private char[] chars;
    private List<String> result;

    private void swap(int i, int j) {
        char temp = chars[i];
        chars[i] = chars[j];
        chars[j] = temp;
    }
    private void dfs(int idx) {
        if (idx == chars.length - 1) {
            result.add(new String(chars));
            return;
        }
        boolean[] visited = new boolean[26];
        visited[chars[idx] - 'a'] = true;
        dfs(idx + 1);
        for (int i = idx + 1; i < chars.length; ++i) {
            if (!visited[chars[i] - 'a']) {
                visited[chars[i] - 'a'] = true;
                swap(idx, i);
                dfs(idx + 1);
                // 回溯
                swap(idx, i);
            }
        }
    }
    public String[] permutation(String s) {
        chars = s.toCharArray();
        result = new LinkedList<>();
        dfs(0);
        return result.toArray(new String[0]);
    }
}

File: python/interview_38.py
class Solution:
    def permutation(self, s: str) -> List[str]:
        chars = [ss for ss in s]
        n = len(chars)
        result = []

        def dfs(idx):
            if idx == n - 1:
                result.append(''.join(chars))
                return
            visited = set()
            visited.add(chars[idx])
            dfs(idx + 1)
            for i in range(idx + 1, n):
                if not chars[i] in visited:
                    visited.add(chars[i])
                    chars[idx], chars[i] = chars[i], chars[idx]
                    dfs(idx + 1)
                    chars[idx], chars[i] = chars[i], chars[idx]
        dfs(0)
        return result
